crates/bevy_ui/src/measurement.rs

```diff
@@ -0,0 +1,83 @@
+use bevy_ecs::prelude::Component;
+use bevy_math::Vec2;
+use bevy_reflect::Reflect;
+use std::fmt::Formatter;
+pub use taffy::style::AvailableSpace;
+
+impl std::fmt::Debug for CalculatedSize {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("CalculatedSize")
+            .field("size", &self.previous_size)
+            .finish()
+    }
+}
+
+/// A `Measure` is used to compute the size of a ui node
+/// when the size of that node is based on its content.
+pub trait Measure: Send + Sync + 'static {
+    /// Calculate the size of the node given the constraints.
+    fn measure(
+        &self,
+        width: Option<f32>,
+        height: Option<f32>,
+        available_width: AvailableSpace,
+        available_height: AvailableSpace,
+    ) -> Vec2;
+
+    /// Clone and box self.
+    fn dyn_clone(&self) -> Box<dyn Measure>;
+}
+
+/// A `FixedMeasure` is a `Measure` that ignores all constraints and
+/// always returns the same size.
+#[derive(Default, Clone)]
+pub struct FixedMeasure {
+    size: Vec2,
+}
+
+impl Measure for FixedMeasure {
+    fn measure(
+        &self,
+        _: Option<f32>,
+        _: Option<f32>,
+        _: AvailableSpace,
+        _: AvailableSpace,
+    ) -> Vec2 {
+        self.size
+    }
+
+    fn dyn_clone(&self) -> Box<dyn Measure> {
+        Box::new(self.clone())
+    }
+}
+
+/// A node with a `CalculatedSize` component is a node where its size
+/// is based on its content.
+#[derive(Component, Reflect)]
+pub struct CalculatedSize {
+    /// Used to track changes, not required to be set.
+    pub previous_size: Vec2,
```

Comment:
Maybe this should be an `Option<Vec2>` and initialized to `None`

Reply:
The calculated size is zero until the image is loaded anyway, so an option doesn't seem useful. This value is just a temporary hack really,  it's here so the image systems don't need need to be rewritten in this PR as well.